cmd/zfs/zfs_project.c

```diff
@@ -0,0 +1,677 @@
+		    strcmp(ent->d_name, "..") == 0)
+			continue;
+
+		if (strlen(ent->d_name) + strlen(pathname) >=
+		    sizeof (fullname) + 1) {
+			ret = -ENAMETOOLONG;
+			break;
+		}
+
+		sprintf(fullname, "%s/%s", pathname, ent->d_name);
+		switch (op) {
+		case ZFS_PROJECT_CHECK:
+			ret = zfs_project_handle_one(fullname, projid,
+			    ZFS_PROJECT_CHECK, hint);
+			break;
+		case ZFS_PROJECT_CLEAR:
+			ret = zfs_project_handle_one(fullname, projid,
+			    ZFS_PROJECT_CLEAR, hint);
+			break;
+		case ZFS_PROJECT_SET:
+			ret = zfs_project_handle_one(fullname, projid,
+			    ZFS_PROJECT_SET, hint);
+			break;
+		}
+
+		if (recursive && ret == 0 && ent->d_type == DT_DIR)
+			ret = zfs_project_item_alloc(head, fullname);
+		errno = 0;
+	}
+
+	if (errno || ret) {
+		if (ret == 0)
+			ret = -errno;
+		(void) fprintf(stderr, gettext("fail to readdir %s: %s\n"),
+		    pathname, strerror(-ret));
+	}
+
+	closedir(dir);
+	return (ret);
+}
+
+static int
+zfs_project_check(const char *pathname, uint64_t projid, boolean_t newline,
+    boolean_t dironly, boolean_t recursive)
+{
+	zfs_project_item_t *zpi;
+	zfsxattr_t fsx;
+	struct stat st;
+	list_t head;
+	int ret, fd;
+
+	ret = stat(pathname, &st);
+	if (ret) {
+		(void) fprintf(stderr, gettext("fail to stat %s: %s\n"),
+		    pathname, strerror(errno));
+		return (ret);
+	}
+
+	if (!S_ISDIR(st.st_mode)) {
+		if (dironly || recursive) {
+			(void) fprintf(stderr,
+			    gettext("'-R' option on non-dir target %s\n"),
+			    pathname);
+			return (-1);
+		}
+
+		ret = zfs_project_handle_one(pathname, projid,
+		    ZFS_PROJECT_CHECK, 1);
+		return (ret);
+	}
+
+	if (dironly) {
+		ret = zfs_project_handle_one(pathname, projid,
+		    ZFS_PROJECT_CHECK, 1);
+		return (ret);
+	}
+
+	if (recursive) {
+		fd = open(pathname, O_RDONLY | O_NOCTTY | O_DIRECTORY);
+		if (fd < 0) {
+			(void) fprintf(stderr, gettext("fail to open %s: %s\n"),
+			    pathname, strerror(errno));
+			return (fd);
+		}
+
+		ret = ioctl(fd, ZFS_IOC_FSGETXATTR, &fsx);
+		close(fd);
+		if (ret) {
+			(void) fprintf(stderr,
+			    gettext("fail to get xattr for %s: %s\n"),
+			    pathname, strerror(errno));
+			return (ret);
+		}
+
+		if (projid == ZFS_INVALID_PROJID)
+			projid = fsx.fsx_projid;
+
+		if (fsx.fsx_projid != projid ||
+		    !(fsx.fsx_xflags & ZFS_PROJINHERIT_FL)) {
+			if (newline) {
+				char *str;
+
+				if (fsx.fsx_xflags & ZFS_PROJINHERIT_FL)
+					str = "P";
+				else
+					str = "[-]";
+
+				if (fsx.fsx_projid != projid)
+					(void) printf("[%u] %s %s\n",
+					    fsx.fsx_projid, str, pathname);
+				else
+					(void) printf("%u %s %s\n",
+					    fsx.fsx_projid, str, pathname);
+			} else {
+				char c = '\0';
+
+				(void) printf("%s%c", pathname, c);
+			}
+		}
+	}
+
+	list_create(&head, sizeof (zfs_project_item_t),
+	    offsetof(zfs_project_item_t, zpi_list));
+	ret = zfs_project_item_alloc(&head, pathname);
+	if (ret)
+		return (ret);
+
+	while ((zpi = list_remove_head(&head)) != NULL) {
```

Comment:
I may be missing something but why can't we do this recursively with a normal depth first traversal?  We're operating in user space and have gobs of stack space.  That should remove the need entirely for the list of directories.

Reply:
In theory, depth first method can do the same work. But one possible advantage for width first method is that: As long as we set project id and inherit flag on the directory, then new created files and subdirectories will keep the same project id and inherit flag. For a large tree, it is normal that someone create new files / subdirectories under the tree that the tree quota is in processing. Width first method will more quickly handle the top level subdirectories firstly, that may save some effort to handle new created files / subdirectories under the tree.